### DISCUSSION TITLE
[General] Passing an array in zksync-cli

### DISCUSSION THREAD

**Discussion by 1am** [zkSync-Community-Hub/zksync-developers#646] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement and the issue you are running into.

Hello

I'm trying to test one contract which requires an address[] as input but it doesn't seem possible with zksync-cli, am I getting it right?
Whatever I try to input it never gets interpreted as array and API encodes it as a s string in [this line](https://github.com/matter-labs/zksync-cli/blob/main/src/commands/contract/read.ts#L126)

My flow looks like below (addresses trimmed for readability, though nothing secret here)

```sh
$ zksync-cli contract read
? Chain to use zkSync Mainnet
? Contract address 0x8....4
? Contract method to call getPairInfo(address[] path, bool stable) view returns (address tokenA, address tokenB, address pair, uint256 reserveA, 
uint256 reserveB, uint256 fee)
? Provide method arguments:
? [1/2] path (address[]) [0x5...E,0x0...A]
>> Failed to encode provided argument: expected array value (argument="path", 
value="[0x5...E,0x0...A]", code=INVALID_ARGUMENT, version=abi/5.7.0)
```

Is there some other way to interact with such contracts, other than writing a script? I find the CLI super useful but got stopped by this situation ;)

**Comment by alhenaai** [comment]
It seems you are encountering an issue with passing an array of addresses as an argument in the ZKsync CLI. The correct way to pass an array of addresses to a method using the ZKsync CLI is to ensure that the array is formatted correctly in the command line interface.

Here’s how you can format the array of addresses:

1. When prompted for the method arguments, you should enter each address in the array separately, ensuring that they are enclosed in square brackets and separated by commas without spaces. For example:
   ```
   [0x5...E,0x0...A]
   ```

2. Make sure there are no spaces after the commas. The CLI might be interpreting the input incorrectly due to formatting issues.

If you continue to experience issues, it might be helpful to check the documentation for any updates or specific instructions on passing array types, or consider using a script for more complex interactions.

For more detailed guidance on using the ZKsync CLI, you can refer to the [ZKsync CLI documentation](https://docs.zksync.io/build/zksync-cli/interact-with-contracts).

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Reply by 1am** [reply]
I've tried this before starting the discussion. From inspection of the sources it seems like the value gets [casted as a string](https://github.com/matter-labs/zksync-cli/blob/main/src/commands/contract/read.ts#L126) and as a result [Array.isArray in ethers.js](https://github.com/ethers-io/ethers.js/blob/v5.7.2/packages/abi/src.ts/coders/array.ts#L195) check fails.

**Reply by alhenaai** [reply]
Thank you for providing additional details and for checking the source code. It appears that the issue is indeed related to how the ZKsync CLI handles array inputs, casting them as strings which then fail the `Array.isArray` check in ethers.js.

Given this situation, the best approach to interact with contracts that require array inputs would be to use a script. This allows you to bypass the CLI's limitations and directly use the ethers.js library, which can handle complex data types more effectively.

Here’s a basic example of how you might set up a script using ethers.js to interact with your contract:

```javascript
const { ethers } = require('ethers');
const provider = new ethers.providers.JsonRpcProvider('https://rpc.zksync.io');

const contractAddress = '0x8....4';
const abi = [
  // Include the ABI of your contract here
];

const contract = new ethers.Contract(contractAddress, abi, provider);

async function main() {
  const path = ['0x5...E', '0x0...A']; // Array of addresses
  const stable = true; // Your boolean value

  try {
    const result = await contract.getPairInfo(path, stable);
    console.log('Result:', result);
  } catch (error) {
    console.error('Error:', error);
  }
}

main();
```

This script sets up a connection to the ZKsync network and interacts with your contract using the `getPairInfo` method. Make sure to replace `'https://rpc.zksync.io'` with the correct RPC URL for ZKsync Mainnet and include the actual ABI of your contract.

For more complex interactions or to automate tasks, scripting provides a flexible and powerful solution.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by uF4No** [comment]
Added as feature request in the zksync-cli repo: https://github.com/matter-labs/zksync-cli/issues/164

**Comment by itsacoyote** [comment, marked as answer]
This has been fixed with https://github.com/matter-labs/zksync-cli/pull/166
